keps/sig-api-machinery/1630-kubeconfig-qps-burst/README.md

```diff
@@ -0,0 +1,130 @@
+- [ ] Design details are appropriately documented
+- [ ] Test plan is in place, giving consideration to SIG Architecture and SIG Testing input
+- [ ] Graduation criteria is in place
+- [ ] "Implementation History" section is up-to-date for milestone
+- [ ] User-facing documentation has been created in [kubernetes/website], for publication to [kubernetes.io]
+- [ ] Supporting documentation e.g., additional design documents, links to mailing list discussions/SIG meetings, relevant PRs/issues, release notes
+
+<!--
+**Note:** This checklist is iterative and should be reviewed and updated every time this enhancement is being considered for a milestone.
+-->
+
+[kubernetes.io]: https://kubernetes.io/
+[kubernetes/enhancements]: https://git.k8s.io/enhancements
+[kubernetes/kubernetes]: https://git.k8s.io/kubernetes
+[kubernetes/website]: https://git.k8s.io/website
+
+## Summary
+
+Kubernetes REST clients use a simple token bucket rate limiter to control the pace of requests from a client.
+We will expose the values already present in the [REST config](https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/client-go/rest/config.go#L114-L120), via a kubeconfig file inside the client stanza.
+We will wire the in-cluster kubeconfig file to honor the presence of a new file to allow customization inside a cluster at the discretion of a different, user provided, controller.
+
+## Motivation
+
+We've always had reasons to override these values, so we created one-off arguments (`--kube-api-burst`, `--kube-api-qps`).
+With widespread adoption of operators and other in-cluster workloads, having a standard way to provide these values makes sense.
+In addition, with the introduction of API priority and fairness, being able to configure and even disable this client-side rate limiting in compatible environments is important.
+
+### Goals
+
+1. Allow easy and consistent configuration for the existing client-side rate limiting.
+
+### Non-Goals
+
+1. Allow customization of client-side rate limiting strategies.
+
+## Proposal
+
+Add to the `AuthInfo` section of a kubeconfig.
+This section identifies a particular client identity.
+This identity spans all the contexts with different namespaces, so it is a closer logical match.
+You don't generally want different QPS characteristics based on namespace.
+
+```go
+type AuthInfo struct {
+	// ...
+	// QPS indicates the maximum QPS to the master from this client.
+	// If it's zero, the created RESTClient will use DefaultQPS: 5
+	QPS float32
+
+	// Maximum burst for throttle.
+	// If it's zero, the created RESTClient will use DefaultBurst: 10.
+	Burst int32
+}
+```
+
+### in-cluster kubeconfig overrides
+The in-cluster config is the automatic configuration that exists within pods.
+The client code itself can be written to consume an overrides.kubeconfig file, which will be merged with the standard
+kubeconfig merging rules that we use the env var path today.
```

Comment:
The merging rules we have today do not allow adding/modifying fields within a user.  The first specification of that user is taken as a whole and no part is overridden.

Reply:
> The merging rules we have today do not allow adding/modifying fields within a user. The first specification of that user is taken as a whole and no part is overridden.

No, we merge all of it:  https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/client-go/tools/clientcmd/loader.go#L226-L252 

You're probably used to selection, but merging is also a thing.

Reply:
Yes, I looked at that code.  You see I raised issues with the poor comment on that function (https://github.com/kubernetes/kubernetes/issues/91092) and the poor user-facing documentation about this (https://github.com/kubernetes/website/issues/20967).

The code in question does not itself answer the question, it just makes a two-handed pile of invocations of `mergo.MergeWithOverwrite`.  The comments on that suggest that merging will be done, but the implementation does not do that.  The implementation does not simply look at each layer of structure independently.  In the case of a map, the behavior depends on the type of each value.  In the case of a map entry whose value is a struct or pointer, the implementation gets to https://github.com/imdario/mergo/blob/v0.3.5/merge.go#L143 --- which replaces the whole value without recursing into it.

I tested, and indeed found that when merging two kubeconfig files that define different fields for a certain user, the result is only the fields of the first kubeconfig file to define that user.